stdlib/public/core/StringStorageBridge.swift

```diff
@@ -335,6 +373,464 @@ extension __SharedStringStorage {
+    mutating func _replace(
+      in range: Range<String.Index>,
+      with replacement: String
+    ) {
+      switch self {
+      case .native(var str):
+        // The next two lines should optimize to an in-place mutation
+        str.replaceSubrange(range, with: replacement)
+        self = .native(str)
+      case .utf16(_):
+        _internalInvariantFailure(
+          "Fast path can't be used on UTF16-backed NSMutableString"
+        )
+      }
+    }
+    
+    // Slow mutation path: replacement may be invalid, we are known Array-backed
+    mutating func _replace_slow(
+      in nsRange: _SwiftNSRange,
+      with utf16Bytes: UnsafeBufferPointer<UTF16.CodeUnit>
+    ) {
+      switch self {
+      case .native(_):
+        _internalInvariantFailure(
+          "Slow path can't be used on String-backed NSMutableString"
+        )
+      case .utf16(var arr):
+        let cocoaRange = nsRange.location ..< nsRange.location + nsRange.length
+        arr.replaceSubrange(cocoaRange, with: utf16Bytes)
+        self = .utf16(arr)
+      }
+    }
+    
+    // Fast path for append
+    mutating func _append(string: String) {
+      switch self {
+      case .native(var str):
+        str.append(string)
+        self = .native(str)
+      case .utf16(var arr):
+        arr.append(contentsOf: string.utf16)
+        self = .utf16(arr)
+      }
+    }
+    
+    // Slow path: bytes may be invalid UTF-16, either representation allowed
+    mutating func _append(utf16Bytes: UnsafeBufferPointer<UTF16.CodeUnit>) {
+      self._switchRepresentationsIfInvalidUTF16(utf16Bytes)
+      
+      switch self {
+      case .native(var str):
+        str.append(String(decoding: utf16Bytes, as: UTF16.self))
+        self = .native(str)
+      case .utf16(var arr):
+        arr.append(contentsOf: utf16Bytes)
+        self = .utf16(arr)
+      }
+    }
+    
+    func mutableCopy() -> AnyObject {
+      switch self {
+      case .native(let str):
+        return _SwiftNSMutableString(str)
+      case .utf16(let arr):
+        return _SwiftNSMutableString(brokenContents: arr)
+      }
+    }
+    
+    func copy() -> AnyObject {
+      // While __StringStorage instances aren't immutable in general,
+      // mutations may only occur when instances are uniquely referenced.
+      // Therefore, it is safe to return self here; any outstanding Objective-C
+      // reference will make the instance non-unique.
+      switch self {
+      case .native(let str):
+        return str._bridgeToObjectiveCImpl()
+      case .utf16(let arr):
+        return _SwiftNSMutableString(brokenContents: arr)
+      }
+    }
+  }
+  
+  private var _contents: Contents
+  
+  internal init(_ str: String) {
+    _contents = .native(str)
+    super.init()
+    _internalInvariant(asString._guts.isFastUTF8)
+  }
+  
+  internal init(brokenContents arr: [UTF16.CodeUnit]) {
+    _contents = .utf16(arr)
+    super.init()
+  }
+  
+  final internal var isStringBacked: Bool {
```

Comment:
Seems like you can replace all those switches with an optional `_tryAsString` or some such and invariant check it not nil.

Reply:
Hmm. The idea being that asString converts, but _tryAsString only does the thing if it doesn't require converting?